cmd/gitlab.go

```diff
@@ -386,86 +362,145 @@ func (c *Client) pollPipelineJobs(p Project, pipelineID int, topics string, ref
+		c.outputPipelineStatusMetric(lastPipeline.Status, p.ShouldOutputSparseStatusMetrics(cfg), gp.PathWithNamespace, topics, ref)
 
-				c.outputPipelineStatusMetric(lastPipeline.Status, p.ShouldOutputSparseStatusMetrics(cfg), gp.PathWithNamespace, topics, ref)
+		if p.ShouldFetchPipelineJobMetrics(cfg) {
+			if err := c.pollPipelineJobs(p, lastPipeline.ID, topics, ref); err != nil {
+				log.Errorf("Could not poll jobs for pipeline %d: %s", lastPipeline.ID, err.Error())
+			}
+		}
 
-				if p.ShouldFetchPipelineJobMetrics(cfg) {
-					if err := c.pollPipelineJobs(p, lastPipeline.ID, topics, ref); err != nil {
-						log.Errorf("Could not poll jobs for pipeline %d: %s", lastPipeline.ID, err.Error())
-					}
-				}
+		timeSinceLastRun.WithLabelValues(gp.PathWithNamespace, topics, ref).Set(time.Since(*lastPipeline.CreatedAt).Round(time.Second).Seconds())
+	}
 
+	return nil
+}
+
+func (c *Client) pollProjects(until <-chan bool) {
+	go func() {
+		pollWildcardsEvery := time.NewTicker(time.Duration(cfg.ProjectsPollingIntervalSeconds) * time.Second)
+		pollRefsEvery := time.NewTicker(time.Duration(cfg.RefsPollingIntervalSeconds) * time.Second)
+		stopWorkers := make(chan struct{})
+		defer close(stopWorkers)
+		// first execution
+		c.discoverWildcards()
+		c.pollWithWorkersUntil(stopWorkers)
+		for {
+			select {
+			case <-until:
+				log.Info("stopping projects polling...")
+				return
+			case <-pollRefsEvery.C:
+				c.pollWithWorkersUntil(stopWorkers)
+			case <-pollWildcardsEvery.C:
+				c.discoverWildcards()
 			}
 		}
+	}()
+}
 
-		if lastPipeline != nil {
-			timeSinceLastRun.WithLabelValues(gp.PathWithNamespace, topics, ref).Set(float64(time.Since(*lastPipeline.CreatedAt).Round(time.Second).Seconds()))
-			b.Reset()
-		}
+func (c *Client) discoverWildcards() {
+	log.Infof("%d wildcard(s) configured for polling", len(cfg.Wildcards))
+	if err := c.pollProjectsFromWildcards(); err != nil {
+		log.Errorf("%s", err.Error())
+	}
+}
 
-		time.Sleep(b.Duration())
+func (c *Client) pollWithWorkersUntil(stopWorkers <-chan struct{}) {
+	log.Infof("%d project(s) configured for polling", len(cfg.Projects))
+	pollErrors := c.pollProjectsWithMaxWorkers(stopWorkers, cfg.Projects...)
+	for _, err := range pollErrors {
+		log.Errorf("%v", err)
 	}
 }
 
-func (c *Client) pollProjects() {
-	log.Infof("%d project(s) and %d wildcard(s) configured", len(cfg.Projects), len(cfg.Wildcards))
-	for _, p := range cfg.Projects {
-		go c.pollProject(p)
+func (c *Client) pollProjectsFromWildcards() error {
+	for _, w := range cfg.Wildcards {
+		foundProjects, err := c.listProjects(&w)
+		if err != nil {
+			return fmt.Errorf("could not list projects for wildcard %v: %v", w, err)
+		}
+		for _, p := range foundProjects {
+			if !projectExists(p) {
+				log.Infof("Found new project: %s", p.Name)
+				//go c.pollProject(p) we don't need to poll projects because they will be picked up at next iteration
```

Comment:
agreed, I think you can even remove this comment 👍

Reply:
thanks removed